architecture/crds.md

```diff
@@ -0,0 +1,324 @@
+
+## The Core Problem
+
+The core problem is that CRDs (being a globally shared resource) are fragile. Once a CRD is installed, we typically have to assume (all other things being equal) that it is shared across namespaces and groups of users.
+
+For that reason, installing, modifying, and deleting CRDs is a process that has ramifications for all users and systems of that cluster.
+
+In spite of this, the Kubernetes API server is permissive about these CRDs. CRDs are mutable (even without a version change). When they are deleted, all instances of them are deleted without warning. They can be upgraded to be entirely incompatible with previous versions. And there is no way to programmatically inspect the CRDs in a cluster and determine whether they are used, how they are used, and by what.
+
+## Users First: A Core Principal
+
+Over the years, several proposals have surfaced and been rejected for one straightforward reason: They did not protect the user from badly written charts. This section explains the reasoning process behind our decision-making.
+
+Helm distinguishes between at least two roles:
+
+- A chart author: A person filling this role _creates_ and _maintains_ Helm charts
+- A Helm user: This person installs and manages instances of charts
+
+### The Chart Author Role
+
+We assume that a _chart author_ has three specific areas of domain knowledge:
+
+1. How to model applications in Kubernetes
+2. How to create Helm charts, including author templates
+3. How the application they are packaging works
+
+In this case, we assume the chart author role includes knowledge of Kubernetes kinds, Helm template commands, and so on.
+
+We do not assume that chart authors have knowledge about the clusters into which their charts are deployed, knowledge of the source code for the packages they install, or knowledge of the extended toolchains that their users have employed. Furthermore, we do not assume that chart authors will always follow best practices or accommodate use cases that may be important to some class of users. In fact, the Helm security model urges us to include bad actors in the class of chart authors. (That is, there is or may be a small subclass of chart authors that have intentions counter to the desires of their target Helm users.)
+
+### The Helm User Role
+
+Our assumptions about _Helm users_ are more modest. We do not assume they know much about Kubernetes or Helm--perhaps only enough to follow the Quickstart guide for Helm. With the base Helm user, we do not assume that they know what a Pod is, let alone a CRD. While we do assume that they know a little about YAML, we make no assumptions that they know about the Kubernetes flavor of YAML.
```

Comment:
When you design software, you must make base-level assumptions about your users. By definition, "base level" isn't "all levels". We can't assume "all levels." That is nonsense. We design with the _assumption that a helm user does not know much_. We don't design with the assumption that some know level X and others know level Y. The matrix you would need to support that design would produce completely unusable software.

Reply:
(In the general case, not being Helm-specific here) the problems with only explicitly addressing the base level of user are:
* It doesn't give any guidance for how to become a more-advanced user.
* It can be used to unreasonably reject features because the base level of user would not know how to use them safely, where it may be a reasonable approach to ensure that the base level of user knows _not_ to use them, analogous to removing the button when what we need is a 'Molly-guard'.

Reply:
Okay, I tried to update that Helm is trying to explain that "at the base level" a Helm user need not have much Kubernetes knowledge. That is, I tried to be explicit that there are of course more advanced users. But that Helm must be accessible to the base-level user.